File: cano-app/api/schedules.ts
import { getAuthHeaders } from './auth';

const API_URL = process.env.EXPO_PUBLIC_API_URL!;

export interface ScheduleDto {
  id: string;
  user_id: string;
  device_id: string;
  name: string;
  action: string;
  payload: Record<string, unknown>;
  cron_expr: string | null;
  next_run_at: string;
  is_active: boolean;
  last_command_id: string | null;
  last_command: { status: string; error: string | null; executed_at: string | null } | null;
  created_at: string | null;
  devices?: { name: string; type: string };
}

export interface CreateScheduleParams {
  device_id: string;
  name: string;
  action: string;
  payload?: Record<string, unknown>;
  run_at?: string;       // one-time: ISO datetime
  cron_expr?: string;    // recurring: cron expression
  timezone?: string;
}

// ── API calls ─────────────────────────────────────────────────────────────────

export async function getSchedules(): Promise<ScheduleDto[]> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/v1/schedules/`, { headers });
  if (!res.ok) throw new Error('Error cargando tareas');
  return res.json();
}

export async function createSchedule(params: CreateScheduleParams): Promise<ScheduleDto> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/v1/schedules/`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: 'Error desconocido' }));
    throw new Error(err.detail ?? 'Error creando tarea');
  }
  return res.json();
}

export async function deleteSchedule(scheduleId: string): Promise<void> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/v1/schedules/${scheduleId}`, {
    method: 'DELETE',
    headers,
  });
  if (!res.ok) throw new Error('Error eliminando tarea');
}

export async function toggleSchedule(scheduleId: string, isActive: boolean): Promise<ScheduleDto> {
  const headers = await getAuthHeaders();
  const res = await fetch(`${API_URL}/api/v1/schedules/${scheduleId}/toggle`, {
    method: 'PATCH',
    headers,
    body: JSON.stringify({ is_active: isActive }),
  });
  if (!res.ok) throw new Error('Error actualizando tarea');
  return res.json();
}

File: backend/app/repositories/schedules.py
from app.core.db import supabase


class ScheduleRepository:

    LIST_SELECT = (
        "*, devices(name, type), "
        "last_command:commands!last_command_id(status, error, executed_at)"
    )

    def find_conflicting_power(
        self,
        device_id: str,
        minute_start_iso: str,
        minute_end_iso: str,
        power_actions: list[str],
    ) -> list[dict]:
        return (
            supabase.table("schedules")
            .select("id")
            .eq("device_id", device_id)
            .eq("is_active", True)
            .in_("action", power_actions)
            .gte("next_run_at", minute_start_iso)
            .lte("next_run_at", minute_end_iso)
            .execute()
            .data
            or []
        )

    def find_by_user_ids(self, user_ids: list[str]) -> list[dict]:
        return (
            supabase.table("schedules")
            .select(self.LIST_SELECT)
            .in_("user_id", user_ids)
            .order("created_at", desc=True)
            .execute()
            .data
            or []
        )

    def find_by_id_and_user(self, schedule_id: str, user_id: str) -> dict | None:
        res = (
            supabase.table("schedules")
            .select("*")
            .eq("id", schedule_id)
            .eq("user_id", user_id)
            .execute()
        )
        return res.data[0] if res.data else None

    def find_cron_meta_in_house(self, schedule_id: str, user_ids: list[str]) -> dict | None:
        res = (
            supabase.table("schedules")
            .select("cron_expr, timezone, user_id")
            .eq("id", schedule_id)
            .in_("user_id", user_ids)
            .execute()
        )
        return res.data[0] if res.data else None

    def find_cron_meta_by_id(self, schedule_id: str) -> dict | None:
        res = (
            supabase.table("schedules")
            .select("cron_expr, timezone")
            .eq("id", schedule_id)
            .execute()
        )
        return res.data[0] if res.data else None

    def find_pending(self, now_iso: str) -> list[dict]:
        return (
            supabase.table("schedules")
            .select("*")
            .lte("next_run_at", now_iso)
            .eq("is_active", True)
            .execute()
            .data
            or []
        )

schedule_repository = ScheduleRepository()
